processing/src/main/java/io/druid/segment/VirtualColumns.java

```diff
@@ -161,10 +162,14 @@ public DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec, Colu
   {
     final VirtualColumn virtualColumn = getVirtualColumn(dimensionSpec.getDimension());
     if (virtualColumn == null) {
-      return dimensionSpec.decorate(NullDimensionSelector.instance());
+      return dimensionSpec.decorate(ConstantDimensionSelector.of(null, dimensionSpec.getExtractionFn()));
     } else {
       final DimensionSelector selector = virtualColumn.makeDimensionSelector(dimensionSpec, factory);
-      return selector == null ? dimensionSpec.decorate(NullDimensionSelector.instance()) : selector;
+      if (selector == null) {
```

Comment:
Shouldn't this logic be inside `virtualColumn.makeDimensionSelector()`? So that `dimensionSpec.decorate()` is applied just at one site?

Reply:
It's only applied if `virtualColumn.makeDimensionSelector` returns null. The contract of the method is that virtualColumn.makeDimensionSelector can return null if a selector can't be made for some reason (maybe the selector type isn't supported).

Reply:
I want `dimensionSpec.decorate()` to be consistently applied at one place. For example, `virtualColumn.makeDimensionSelector()` accepted `ExtractionFn`, and then `dimensionSpec.decorate()` applied here to either result of `virtualColumn.makeDimensionSelector()` call or `ConstantDimensionSelector.of(null, dimensionSpec.getExtractionFn())`.

Reply:
The idea behind the contract of `virtualColumn.makeDimensionSelector` is that it's responsible for applying decoration, and that may give it the power to potentially optimize something in some way (maybe it's aware of certain kinds of DimensionSpecs). Currently, no implementation in core actually uses this power, so it could be taken away if we want and only applied "outside". But the idea was that maybe someone would find a creative use for it.

What do you think?

Reply:
Ok, then returning to the original suggestion, why couldn't `virtualColumn.makeDimensionSelector()` return NullDimensionSelector itself (never returns null)?

Reply:
It could, but then we will still need to apply decoration in two places, since `VirtualColumns.makeDimensionSelector` might still be called with a column name that doesn't exist, and it would have to build and decorate its own column.

However, now that I think about it, we could forbid users from calling `VirtualColumns.makeDimensionSelector` with a column that doesn't exist. No current callers do it anyway. So I can make that change, and make `VirtualColumn.makeDimensionSelector` non-nullable, if that sounds good.